osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1297 @@
+
+  case IPMI_RAW_VALUE_PRESENT:
+    r["value"] = std::to_string(raw_value);
+    break;
+
+  default:
+    r["value"] = std::to_string(val) + getSensorThresholdSuffix(sensor);
+  }
+
+  r["threshold_out_of_range"] =
+      (ipmi_is_threshold_out_of_range(states, IPMI_LOWER_NON_CRITICAL) ||
+       ipmi_is_threshold_out_of_range(states, IPMI_UPPER_NON_CRITICAL))
+          ? "1"
+          : "0";
+
+  IPMIClient* c = reinterpret_cast<IPMIClient*>(data);
+  c->insertRowsQueue(r["name"] + "-" + r["mc_name"] + "-" + r["mc_id"], r);
+}
+
+void getThresholdSensorCB(ipmi_entity_t* entity, void* data) {
+  ipmi_entity_iterate_sensors(
+      entity,
+      [](ipmi_entity_t* ent, ipmi_sensor_t* sensor, void* data) {
+        if (ipmi_sensor_get_event_reading_type(sensor) ==
+            IPMI_EVENT_READING_TYPE_THRESHOLD) {
+          int rv = ipmi_sensor_get_reading(sensor, readThresholdSensorCB, data);
+          if (rv != 0) {
+            LOG(ERROR) << "Could not get sensor reading: " << strerror(rv);
+          }
+        }
+      },
+      data);
+}
+
+/// Walks FRU nodes and appends to Row.
+void traverseFRUNodeTree(ipmi_fru_node_t* node, Row& row) {
+  time_t tm;
+  enum ipmi_fru_data_type_e dtype;
+  double floatval = 0;
+  int intval = 0;
+  unsigned int dataLen = 0;
+  const char* name = nullptr;
+
+  std::unique_ptr<ipmi_fru_node_t, std::function<void(ipmi_fru_node_t*)>> sNode(
+      node, ipmi_fru_put_node);
+
+  for (size_t i = 0;; i++) {
+    char* data = nullptr;
+
+    ipmi_fru_node_t* subnode = nullptr;
+    int rv = ipmi_fru_node_get_field(node,
+                                     i,
+                                     &name,
+                                     &dtype,
+                                     &intval,
+                                     &tm,
+                                     &floatval,
+                                     &data,
+                                     &dataLen,
+                                     &subnode);
+    if (rv == EINVAL) {
+      break;
+    }
+
+    std::unique_ptr<char, std::function<void(char*)>> sData(data, [](char* d) {
+      if (d) {
+        ipmi_fru_data_free(d);
+      }
+    });
+
+    if (rv != 0) {
+      continue;
+    }
+
+    std::string colName;
+    if (name == nullptr) {
+      colName = "missing[" + std::to_string(i) + "]";
+    } else {
+      colName = name;
+    }
+
+    switch (dtype) {
+    case IPMI_FRU_DATA_INT:
+      row[colName] = INTEGER(intval);
+      break;
+
+    case IPMI_FRU_DATA_TIME:
+      row[colName] = BIGINT(tm);
+      break;
+
+    case IPMI_FRU_DATA_ASCII:
+      row[colName] = data;
+      break;
+
+    case IPMI_FRU_DATA_BOOLEAN:
+      row[colName] = INTEGER(intval);
+      break;
+
+    case IPMI_FRU_DATA_FLOAT:
+      row[colName] = std::to_string(floatval);
+      break;
+
+    case IPMI_FRU_DATA_SUB_NODE:
+      traverseFRUNodeTree(subnode, row);
+      break;
+
+    default:
+      // we don't want any of the other data
```

Comment:
perhaps here you could set `row[colName] = "-1";`? This value typically denotes an error of some form.